### ISSUE TITLE
twinBASIC Roadmap / Priorities

### ISSUE THREAD

**Issue by WaynePhillipsEA** [WaynePhillipsEA/twinbasic#75]
There are some big things planned with twinBASIC.  I'd like to hear from as many of you as possible, to hear what **you** think we should be concentrating on in the near future. 

Let's only discuss the bigger issues here.  The smaller issues will be tackled ad-hoc.  

The issues below will each take >1 week to implement.   Please list the following issues in order of priority, **for you**:

***

**Finalize 64-bit support** 

- ~2-3 weeks work.

**Inheritance**     #73

- ~2 weeks work
- Proper class Inheritance support, allowing for full fledged OOP goodness. 

**GUI**   

- GUI support and designers. This will likely take several months to get something remotely usable, with decent VB6-compatibility taking 6-12 months.

**Cross-Platform**   

- Initially we'd look at just x86/x64 support on Linux / Mac.  We'd start with a limited feature-set and then expand over time.   Probably 2-months to get the basics sorted, allowing for generating Linux/Mac executables with a limited feature-set.

**Optimized compiling**   

- ~4-6 weeks
- Using LLVM, generate optimized binaries.

**Analysis Extension API**   #51

- ~2-4 weeks
- Allowing for extensions to be written for twinBASIC that provide custom compile-time inspections/warnings  

**Multi-threading language support**

- ~4-6 weeks
- Whilst twinBASIC does support multi-threading internally, this is not currently exposed via proper language features  

**Edit & Continue support**    #28

- ~3-4 weeks
- Self-explanatory.   Lots of edge cases.

**Comment by Kr00l** [comment]
I would compare this like a building construction. What would make sense to work first with biggest surplus (less work) on other things and to not touch things twice?
That was my say. :-)

**Comment by mansellan** [comment]
I would say that the first priority should be to reach feature-parity with VB6, then build from there. So the GUI side should be near the top. It may be possible to split it up though - once [#62: External Files](https://github.com/WaynePhillipsEA/twinbasic/issues/62) is done, twinBASIC should be able to serialise/deserialise both the `.frm` and `.frx` formats (and equivalents for other component types), next up should be to be able to display UI components at runtime. The designer, whilst important, could be tackled last - in the interim, changes can be made by modifying the module header directly.

**Comment by Greedquest** [comment]
You should pin this issue:)

**Comment by bclothier** [comment]
0. 64-bit support
1. Multi-threading
2. Analysis Extensions
3. Inheritance
4. GUI
5. Cross platform

My reasoning works as follow:

Right now there is little to no value in getting a linux user to adopt tB over alternatives given the handicap of not having any GUI elements or long coattail of a COM ecosystem to work off. For that reason, I think tB needs to mature considerably before it can start offering non-Windows users some value.

However, you may have noticed that I put other stuff ahead of the GUI. Why? Because:
1) GUI necessarily needs multi-threading. In VB6, this was achieved by hiding the implementation details. But with issues like #39 , that needs to be addressed face-on. 
2) I rate the inheritance and analysis higher because this has huge potential to enable those with legacy codebase to get it cleaned up and fix issues, including project-specific issues that the maintainers can define for themselves. Since the goal of tB is to help ease the adoption of codebase from VB6/VBA, it need to offer value in enabling analysis and safely fixing. We do want to bring forward the code but NOT the mentality of "this has been working for last 30 years, don't touch it!"
3) 64-bit support is your "instant selling proposition" because simply migrating a VB* project to tB and suddenly have it run faster _and_ work on 64-bit would be a huge boost. For that reason it's the number 1 on the list, even though the GUI parts is needed to make it even more beneficial.

**Comment by Kr00l** [comment]
I agree with bclothier that cross-platform can be a long-term goal as the VB6/VBA folks are anyway used to rely on windows and there would be nearly no benefit. Thus that point should be the last.

**Comment by A9G-Data-Droid** [comment]
I would like the ability to use a GUI elements (.frm & .frx) exported from a regular forms designer, like VB6 or the VBE. My main use case for tB right now is extracting elements from VBA forms and compiling them to be stand alone so they no longer need the host application to run. The forms designer can come later because for now I can use the forms designer I have. I think that multi-threading should follow right after GUI for the reasons bclothier pointed out. We are going to need to run code on background operations while keeping the GUI responsive.

I think that getting 64bit and GUI support out first will be the selling point. Money will then roll in to help fund the rest of development.

0. 64-bit support
1. GUI    
2. Multi-threading
3. Analysis Extensions
4. Optimized compiling
5. Inheritance
6. Cross platform

**Comment by Kr00l** [comment]
For completeness, here my vote:

0 64-bit support
1 Optimized compiling
2 GUI
3 Edit & Continue support
4 Multi-threading
5 Inheritance
6 Analysis Extension
7 Cross-Platform

**Comment by EduardoVB** [comment]
I think you should analyze the whole project and decide what would be a better and a more logical path from the technical point of view.
Something more or less like what Kr00l said in his first message (about the building construction).

But if you want to decide or at least take into account what would be best for marketing, I guess then that this is the purpose of this discussion/survey. Then I'll give the list of features in order of importance for me, that can be different from the importance that each features has for other developers (and of course, it might not be the best choice from a architectural point of view, and nor even for a marketing point of view).

1. Edit & Continue
2. GUI
3. Optimised compiling
4. 64-bit
5. Multi-threading
6. Cross-platform
7. Inheritance
8. Analysis Extension

I would keep clear and in sight the goal: being a VB6 replacement with backward compatibility.
So all other desirable features like Analysis Extension, Inheritance, Cross-platform, Multi-threading and even 64-bit compiling IMO should be part of a second goal.

I don't think that 64-bit compiling is actually very important technically now, because 32 bits programs work currently well, and with little performance penalty (most of the times). But it can become important if Windows lower support for 32 bits programs at some point.
But from a marketing point of view, yes, it is important. And perhaps not too complex to achieve comparing to others (you said 2-3 weeks).

**Comment by mansellan** [comment]
64-bit has a very important use-case - compiling libraries for use by 64-bit VBA hosts (such as Office).

**Comment by Greedquest** [comment]
Based purely on what I think is most useful/exciting for me:
 1. 64-bit (execution, not necessarily compilation or development). I say with a caveat, since I _think_ if #61 is sorted, then I'll be able to run 32-bit target code from 64 bit office using a COM surrogate app like [this](https://stackoverflow.com/a/42581513/6609896) so I'll at least be able to test the tB code from VBA.
 2. GUI loading (i.e. run form code and display forms at runtime, but not edit/create them in the editor yet)
 3. Multithreading language support
 4. Inheritance
 5. Optimised Compiling
 6. Edit-Continue
 7. Analysis Extension
 8. GUI part 2, designer/editor (jumping between Excel form editor and tB doesn't bother me)
 9. Cross-Platform

However putting myself in your shoes as others have, I don't know if I'd stick to that order. I am only one of 4(ish) categories of VB* user I can think of (well, grouping by use-case, you could choose to group in other ways):
 1. Sunk Cost users - "If we were doing this all again VB6 wouldn't be the number one choice but we're stuck with it".
 2. Misguided individuals who use VB* for the challenge and intellectual curiosity.
 3. VBA has a large niche; integrated with MS Office (and other apps) makes it actually the best tool for the job in some cases.
 4. Possibly VBA/VB6 have smaller niches where they are the best tool, but I can't think of any.



---

These 4 groups derive different benefits from the features you propose; for a few groups a feature is an active draw/exciting, for others it's a case of more-is-better, and for some the feature is a must-have. E.g.

 - 64-bit execution is a real draw for the sunk-cost group, as it reduces the feature gap between what they have and the language they'd rather use. It will also grow the size of the group for which VB* is the best choice - more niches will be accessible. For office integration users, it's probably unimportant as their uncompiled VBA runs fine on 64 bit hosts - 64-bit library support though would make tB valuable. For the enthusiasts, it's nice to have.
 - Full GUI support is a must-have for some legacy users who need at least VB6 parity. More-is-better for most
 - Optimised compiling is not a must-have for anyone (except new niche use cases), since tB is probably as fast as compiled VB6 already.
 - Multithreading - unlikely to be a must-have for anyone except the group 4 niche users "Justified VB* is better than python or c# for this particular use case".
 - etc.

<sub>
<details>
<summary>In UX this is the Kano model</summary>


![image](https://user-images.githubusercontent.com/25348920/115571011-7f64eb80-a2b6-11eb-8057-2aa29888adbf.png)

</details>
</sub>

If you are trying to grow into one of these groups then you will have to develop those must-have features before you can get them to migrate to/ integrate tB. However not all groups have the same needs so whatever you develop and in whatever order it will probably benefit someone.

So I'd recommend developing features:
 - In the order that benefits the group(s) you'd like to prioritise
 - In a sustainable way which won't lead to burn out
 - Focussing on what motivates you:
 	- Cash flow with features that benefit groups you can charge and whatever associated must-haves
 	- Engagement from the users you already have
 	- Unicorns

Hope that helps:)

**Comment by WaynePhillipsEA** [comment]
Some fantastic replies!   Thanks to everyone that has responded so far.  Your thoughts and views are being listened to :)

**Comment by mansellan** [comment]
Can I make a suggestion? Once you've decided on the best priority order, consider opening a GH project so that you can drag the issues into priority order, Kanban-style.

**Comment by WaynePhillipsEA** [comment]
@mansellan Will do, thanks.

**Comment by wopag** [comment]
GUI (load and use existing forms from original source)  ( as per  @mansellan  )
    GUI (form editor)
    64-bit support
    Inheritance
    Analysis Extensions
    Multi-threading   
    Cross platform

Some really good valid points in the suggestions above but im going to give my suggestions purely based on my personal reasons.     I love the simplicity of VB, if I cant more complicated I goto C#,   .NET was a bag of nails in my opinion - hence I only use it when I have no choice. 

I would love an open of a VB6 program and it just works with no intervention,  sure the extensions that you have/planned are all excellent but are future plans for me personally.  I push and pull VB6 now to make it do what I want it todo and if it doesn't I fall back to other languages.  So the "extended" features I feel are not essential to me.

After a few hours of "testing" the preview its all *very* nice and indeed comfortable :)

**Comment by mwolfe02** [comment]
This is mostly @mansellan's list, but I've moved GUI support lower because it's not important to me, personally.  As for the good of the twinBASIC project, I think it probably belongs close to the top of the list:

1. 64-bit
2. Multi-threading
3. Edit & Continue
4. GUI (at least runtime)
5. Inheritance
6. Optimised compiling
7. Analysis Extension
8. Cross-platform

I think 64-bit is something that VBA developers are going to start caring about a lot more now that Microsoft recommends installing Office as 64-bit.  Up through Office 2016, the default install mode was 32-bit.  That switched to 64-bit with the Office 2019 release.

As for multi-threading, I see that as a big draw for Access developers who could throw long-running database operations into a separate thread as that tends to be the biggest performance bottleneck in data-heavy applications.

**Comment by mansellan** [comment]
Also important to note that 64-bit is the point at which twinBASIC goes commercial - it's important that the developers get paid for the long-term success of the project. I'm looking forward to being able to buy a license.

**Comment by dmrvb** [comment]
My interest is VB6 rather VBA so the GUI side is essential, I write software for Windows and everything I do has a GUI. VB6 has two parts, the V and the B, you have almost cracked the B but without the V it just isn't VB .
I can see its difficult, its a big job and a form designer for VScode is much wider than just VB, somebody else just might be doing it so your work might be wasted?
Just do the run time side for now??. I really want to use your Longlong but once I do that I can no longer use VB6 IDE, so will have to code in TwinBasic and do the forms in VB6, I can manage that, but aim to make it as easy as you can.
(if you do the runtime side maybe that will inspire others to do a VScode form designer???)

64 bit is pretty essential as it puts VB right up to date, even though loads of software is still 32 bit.

Inheritance? Ive managed without since VB6 came out so it can't be that important.
Better compiler? I need to run some tests on your current one before I can comment
Edit and Continue is one of the lovely features of VB so should be quite high priority.
Cross platform would be great but should only come along if and when everything else is sorted out. Android and iPhone/iPad are the two big markets. In the real world nobody uses Linux?? Moving VB from desktop to phones is a big step.

**Comment by Kr00l** [comment]
Out of curiosity, how is it planned that the multi-threading language support look like?
I wonder how the syntax will look like and how the main thread can then monitor if a new thread is still running or ran to completion already.
Or will it be an in-built class rather than language keyword? Such a class could hold other information as well, such as hThread etc.

**Comment by xyigor73** [comment]
1) Inheritance
2) Multi-threading
3) Finalize 64 bit
4) GUI
5) Optimized compiling
6) Cross-platform
7) Edit & Continue

I put inheritance first. I believe there must be a complete OOP management.
Multithreading in second place because it is essential to have BackgroundWorkers, for example in the graphical interface to make it fluid and responsive.
These two characteristics are the main ones that I consider essential for the development of complex projects.

64 Bit has to follow, all this work has to be funded.

**Comment by adakordev** [comment]
GUI, import Excel/ Access made forms as a first step
better JSON support
ArrayList, Maps
If you have compiled a twinBASIC .exe, is it going to be possible to store some simple values over sessions?

In my current company VBA is deactivated, but twinBASIC seems to work. Nice.

**Comment by AstrocalcVB** [comment]
This is 7edm from VBF and this is my priority list, which I will rant on below. 
 
0. Inheritance
1. Edit & Continue
2. 64-bit
3. GUI
4. Optimized compiling
5. Multi-threading
6. Cross platform

First of all, my view on this is that the 100% VB6 compatibility goal, of course is of highest importance, but maybe also an ongoing process that embraces almost everything as it's not just a state to reach but also to keep. So, it doesn't really have a place in the list but inspirits it.

Then, while it would be, will be, a dream come through to open a new modern GUI and load my old VB6 ( .vbp) projects, with an estimated 6-12 months to have something useful, I think there are a few useful things that will take shorter time that can be completed/implemented first. 

While inheritance and full OOP surely falls outside of the 100% VB6 objective, it's something I (and I know many with me) really have missed in VB6 for such a long time, and if it only will take 1-2 weeks... with that in place I would have something meaningful to do (redesigning my classes) while waiting for the GUI to take form. 

I am aware that Inheritance has grown out of fashion somewhat lately, but there are many use-cases where it is optimal, and there are others when composition may be a better solution - there is never one solution/technique that fits all - but with OOP in place, in whatever form, I think it really would boost tB's temperature in not only the Basic but also other programing families.

Being able to debug and edit code and then continue, would of course remove a lot of possibly annoying moments, and also falls within the VB6 100% objective. A small but important detail.

I don't really feel an immediate need for 64 bits, and it also falls outside of the VB6 100% objective, but I moved it higher in my list because it also comes with support for this project. So even if I myself wont hurry there, it feels absolute mandatory that this project wont fail, for one or the other reason.

If the GUI will take 6-12 months reaching a useful state, it may be a good idea if it's development and progress took place alongside other things, just an idea. Well, that will be the day when it arrives, or even the first taste of it, and emotionally it's no doubt no. 1 on the list, but the emotional is usually not the most practical approach.

Optimized compiling and Multi-threading are all good and welcomed, but doesn't appear that urgent at this stage.

As for cross platform, I'm not sure if there is an ambition for tB itself to run on Mac and Linux? If so, I'm a bit skeptical to if it's worth the trouble, if there is an audience big enough making it worth it... Being able to create binaries and full set programs to be run-able there is a different matter. It doesn't feel urgent though, and would probably benefit from a later implementation when tB has matured for a while. 

And, I'm really excited about this project!

**Comment by WindowStations** [comment]
Does Twin Basic currently support User Controls created with VB.NET and Microsoft's Interop toolkit templates?  These controls currently work on VB6 and VBA, and they behave like VB.NET controls with a full property list to take advantage of.  I strongly recommend this kind of support to achieve 100 percent (virtually) compatibility between VB6/VBA/VB.NET.  Furthermore, components can be extend with vb.net namespaces to be consumed by vb6/vba/TB

**Comment by A9G-Data-Droid** [comment]
@WindowStations It looks like MicroSoft took down the download pages for the Microsoft's Interop Toolkit. Where are you getting it? Is there a secret mirror somewhere to keep it alive?

**Comment by bryceschober** [comment]
FWIW, I got a ~2.6MB download from https://microsoft-interop-forms-toolkit.software.informer.com/download/

**Comment by A9G-Data-Droid** [comment]
@bryceschober I found it at archive.org, which is a mirror. I think this is safer than your link: https://web.archive.org/web/20200810070025if_/https://download.microsoft.com/download/3/7/D/37D90824-1BE8-4085-920E-54812C045A55/InteropFormToolsInstaller.msi

**Comment by WindowStations** [comment]
Oh yes, the link was alive in 2019-ish, but they have taken it down.  I saved my copy.  The good news is that the compiled dll doesn't need the kit to work on a target machine.  Those both look good.  I have version 2.0a and 2.1, but preferring 2.0a actually since I'm using VS2008 to compile.   I uploaded a video to show that it is possible to get VB6 to accept nested class buckets with syntax This.That.Something.Else.Here("").  Video link: https://www.youtube.com/watch?v=N9QAgIpwS5A

**Comment by mansellan** [comment]
Just my 2c on class inheritance:

It's a tool to be considered, sometimes optimal, sometimes not. It represents an "is-a" relationship, rather than a "has-a" relationship for interfaces. Sometimes composition is better, sometimes pure functions / delegates are better. But sometimes, class inheritance is the most obvious, natural, and expressive way to go.

VB6 is "kinda OOP", in that it does interface inheritance. It's polymorphic, but without some of the modern conveniences. That can lead to boilerplate if you want to write common OOP patterns in VB6.

If twinBASIC can fill in some gaps here, whilst honoring its roots, that could allow us to choose the best way to express intent.

**Comment by retailcoder** [comment]
As someone promoting full-blown OOP in VBA/VB6 (with Rubberduck anyway... you'd be nuts to do this without RD), in my opinion inheritance in twinBASIC is 100% a must-have, no questions, no ifs, no buts. Sure composition can work around the lack of it - to a degree...

Composition worked relatively well for my OOP/MVC Battleship's AI implementations, but my OOP/MVVM proof-of-concept library desperately needs proper inheritance. @mansellan "can lead to boilerplate" is a massive understatement!

**Comment by WindowStations** [comment]
Wayne, can you answer my question about supporting interop controls and namespaces on TB?  The video I posted above shows vb.net namespaces consumed by VB6 as nested class buckets (a term used by MS before NET).  See the last minute of the video for namespaces.  There are a few tricks that can share about that if you are interested, and have the time to work together.  Otherwise, if I don't get any responses then I will drop support.  Hoping to hear from you soon, thanks.

**Comment by EduardoVB** [comment]
If it is supported by VB6 then it has to be supported by tB. I think he has already stated that it will be backward compatible.

**Comment by WindowStations** [comment]
Right, I know that the aim for TB is to be backwards compatible, but I'm seeing things coming up short at the moment.  I'm not sure he is aware of the interop kit, otherwise he'd (and others)  would be using it already.   Also, I don't know how he's implementing comvisible components so we can't assume it will "just work".  Can anyone using TB and the interop kit ( bryceschober or A9G-Data-Droid ) confirm?  A simple default user control should be referenceable without a full designer.  In fact, you can create a form designer as a customized interop user control, using your own custom controls instead.  I just figured that out today, so I know its a solution that is currently available and will save development time, especially if Wayne is thinking in the wrong direction.  I know he doesn't like constructive criticism in a social hierarchy but, a spade is a spade, I'm trying to help.

**Comment by Kr00l** [comment]
It's a preview, keep cool!!

**Comment by EduardoVB** [comment]
AFAIK, it is an alpha version, and it currently supports bas and class modules but not forms and controls, so there is no way you could place controls on a form via interop or any other means (in the current alpha version) because there are still no forms.
You can ask this feature to be supported in the final version (to be able to add controls via interop), but I think it is already covered and answered because it aims for backward compatibility with VB6, so if it works now in VB6 it should also work in the final version of tB.
That final version of tB is still not too close in time, so if you are interested, you need to be patient.
That's how I see it.

**Comment by bclothier** [comment]
> I know he doesn't like constructive criticism in a social hierarchy but, a spade is a spade, I'm trying to help.

I'm sorry to hear that it is hard for you to take constructive criticism in a social hierarchy. I am confident with some practice you'll get better at taking them. One thing that really helps is to focus on the issues, not personalities. 

My advice would be to actually try and reference the assembly. As long they are COM-visible, they should in theory work. If they don't, open a separate issue about this topic specifically. Given this thread is to suggest the priority for the roadmap (see the OP), I would suggest that you keep the discussion in a new thread should you encounter an issue with using the COM-visible assembly built using interop toolkit. 

That said, while this might be an acceptable short-term workaround for some, the dependency on .NET assemblies is probably not the end goal.

**Comment by WindowStations** [comment]
Right the roadmap is important from the OP.  "Let's only discuss the bigger issues here" - "taking 6-12 months"  Perhaps the end goad is for cross platform (another big issue/decision), without net assembly dependency.  However, it will not take long to create a good control set and form designer for Windows support.  It may be necessary to split and avoid a one size fits all approach, but Idk the specifics of the devs point of view.  Plus, more people would be interested to back the project, if there were a shiny UI.  Even though the developer might know that other things are an important priority to get done first.

**Comment by WaynePhillipsEA** [comment]
@WindowStations, I agree with the others replies to your questions above.   I think the 100% backwards compatibility goal is quite clear, so please give what you propose a try and report any issues.  If we do then "come up short", perhaps you'll kindly let us know in what ways we can improve?

We will not be **relying** on any .NET technologies for GUI support in twinBASIC.  Sorry, not happening.   .NET is not what this project is about.

With regards namespaces, the plan is to provide proper namespace support, not class based simulation of some aspects of them (See #111).

**Comment by WindowStations** [comment]
Alright, I will provide a single sample control with source code.  Give me a few days to post it up.  I think it should work out of the box, but i's a good idea to confirm this earlier than later.

I'm not sure you'd be relying on .net tech, you'd be supporting the existing legacy with a feature set that you will never be able to duplicate on your own as a single developer.  Same with Namespace support.  Is you're goal is to do the UI support before cross platform?  Or, maybe these things would unfold simultaneously?   You're providing VB.NET syntax enhancements as a major appeal of your project, so I'd say that the IDE is "about" .NET styling.  I understand the urge to reject dependency on NET, but that's limiting the potential on Windows systems.

**Comment by EduardoVB** [comment]
Why don't you use .NET? It will be a shorter path to you.

**Comment by WindowStations** [comment]
EduardoVB, not at all.  I've programmed in Vb.NET and published articles for years, but always kept a VB6 version to compare efficiency and longevity.  So I'm very well versed in both incarnations, especially the conversions between them (back and forth).  The NET IDE's are over bloated (especially 2010+), very sluggish, and the frameworks have been inconsistent over the years, ie Cannot install framework x because it is already installed with Windows, but the app requires a very specific version that you built it with.  This requires you update the application and isntaller/checker, in many cases for no good reason.  Every few years your project fails to install on newer machines.  There are many other issues with that paradigm.  However, there are many good features that exist to be ported back to VB6 which is still superior in many ways.  Not all ways, but the gap is closing rapidly when you rebuild and update the IDE with modern features.

**Comment by WindowStations** [comment]
Ok, I whipped up an example that should work.  I tested it on my fresh install of Windows 10.  The project sources files are included along with the compiled files, if you trust the author of course.  The VB.NET project was compiled in VS2008 architecture edition, called WidgetControls.  It provides a VB.NET button control name ButtonControl, with advanced NET properties, visual styles, and events.  Also, a I included a second control as a basic starter template for a form designer.  It's called FormDesignerControl, and it should be referenceable by TB as is.  No properties or events are included in the demo, so I'm not using the WithEvents keyword.  It would take several hours or days to complete something remotely useable, perhaps a month or two to perfect and test.  The kit does offer backwards compatibility with VB6 Forms, so have at it...  I also included the project to help register and unregister NET dll assemblies.  Simply drag and drop the NET dll into the executable SetupNETAssembly.exe and SetupUnregisterAssembly.exe respectively.  Link to the github sources here: 

https://github.com/WindowStations/WidgetControlsDemo

The original C# source code for the designer came from the Code Project website, although don't use it because it's broken and comes up as a virus by MS defender.  I updated the source to run in VB.NET.  The problem was occurring when a control was added to the form with Me.Controls.Add.  You must use Me.Controls.AddRange instead for some reason.  It works now!  https://www.codeproject.com/Articles/36190/Design-Time-Architecture-in-NET-2-0-Part-1

**Comment by AstrocalcVB** [comment]
@WindowStations  Please open a new issue about this giving it the space to get proper attention. The post above looks like perfect start. In my view, you have already expressed your vote of priority for "interop controls and namespaces" and Wayne has answered.  What you post here now is simply what needs to go in a new issue. I'm not sure how github works here, but I think you can open it directly from the ... menu of your post and create a reference link to that issue here. Everything would be kept clean and tidy., everyone happy and in productive mode.

**Comment by WindowStations** [comment]
Thanks, but Wayne said: "so please give what you propose a try and report any issues".  I've made the proposal and sources available, but I don't have any confirmed issues to report yet.  I uninstalled TB, so I'm asking someone to test this to see if there is any issue to open.  If it works, then we can be more productive by moving on.  It should work.

**Comment by KarlSevenSeven** [comment]
Me:

1. VB6 compatibility
2. VB6 compatibility incl. forms
3. Optimized compiling
4. Edit & continue
5. 64bit
6. Multi-threading
7. GUI designer


to 1.
Loading VBP and run/compile with _very minor_ changes or better _no_ changes.

to 2.
Like 1. including _existing_ forms.
We have a form designer which is cumbersome but works...

**Comment by WindowStations** [comment]
Is the preview (version 1) continually updated with the newest features of the update log? 
Has anyone been able reference (dll/tlb) NET assemblies as a com visible wrapper? 

 I would like to test TB with VB6Porter and VB6Namespaces when this feature is ready.  Thanks

**Comment by WaynePhillipsEA** [comment]
@WindowStations Yes, it's continually updated, and should auto-update as long as you haven't changed the auto-update settings in VS Code.

When you open VS Code, check the version number by opening the Extensions panel:

<img width="291" alt="tb_CheckVersion" src="https://user-images.githubusercontent.com/25738109/128295401-0aaa8bea-e17e-4e65-8011-aee5779164fb.png">

As of today, we're on v0.10.3900
